frontend/components/markup-modal/markup-modal-content.js

```diff
@@ -0,0 +1,177 @@
+    // https://github.com/badges/shields/blob/master/services/service-definitions.js
+    example: PropTypes.object,
+    baseUrl: PropTypes.string.isRequired,
+  }
+
+  indicatorRef = React.createRef()
+
+  state = {
+    path: '',
+    link: '',
+    markupFormat: 'link',
+    message: undefined,
+  }
+
+  get baseUrl() {
+    const { baseUrl } = this.props
+    if (baseUrl) {
+      return baseUrl
+    } else {
+      // Default to the current hostname for when there is no `BASE_URL` set
+      // at build time (as in most PaaS deploys).
+      const { protocol, hostname } = window.location
+      return `${protocol}//${hostname}`
+    }
+  }
+
+  generateBuiltBadgeUrl() {
+    const { baseUrl } = this
+    const { path, queryString } = this.state
+
+    const suffix = queryString ? `?${queryString}` : ''
+    return `${baseUrl}${path}.svg${suffix}`
+  }
+
+  renderLivePreview() {
+    // There are some usability issues here. It would be better if the message
+    // changed from a validation error to a loading message once the
+    // parameters were filled in, and also switched back to loading when the
+    // parameters changed.
+    const { baseUrl } = this.props
+    const { pathIsComplete } = this.state
+    let src
+    if (pathIsComplete) {
+      src = this.generateBuiltBadgeUrl()
+    } else {
+      src = staticBadgeUrl(
+        baseUrl,
+        'preview',
+        'some parameters missing',
+        'lightgray'
+      )
+    }
+    return (
+      <p>
+        <Badge display="block" src={src} />
+      </p>
+    )
+  }
+
+  copyMarkup = async markupFormat => {
+    const {
+      example: {
+        example: { title },
+      },
+    } = this.props
+    const { link } = this.state
+
+    const builtBadgeUrl = this.generateBuiltBadgeUrl()
+    const markup = generateMarkup({
+      badgeUrl: builtBadgeUrl,
+      link,
+      title,
+      markupFormat,
+    })
+
+    try {
+      await clipboardCopy(markup)
+    } catch (e) {
+      this.setState({
+        message: 'Copy failed',
+        markup,
+      })
+      return
+    }
+
+    this.setState({ markup })
+    this.indicatorRef.current.trigger()
+  }
+
+  renderMarkupAndLivePreview() {
+    const { indicatorRef } = this
+    const { markup, message, pathIsComplete } = this.state
+
+    return (
+      <div>
+        {this.renderLivePreview()}
+        <CopiedContentIndicator ref={indicatorRef} copiedContent="Copied">
+          <RequestMarkupButtom
+            isDisabled={!pathIsComplete}
+            onMarkupRequested={this.copyMarkup}
+          />
+        </CopiedContentIndicator>
+        {message && (
+          <div>
+            <p>{message}</p>
+            <p>Markup: {markup}</p>
+          </div>
+        )}
+      </div>
+    )
+  }
+
+  renderDocumentation() {
+    const {
+      example: { documentation },
+    } = this.props
+
+    return documentation ? (
+      <Documentation dangerouslySetInnerHTML={{ __html: documentation }} />
```

Comment:
This is fine, but usually with `dangerouslySetInnerHTML` you'll want to wrap the HTML in a `{__html}` object at the point where the HTML is created rather than in the component itself, and then pass around that object rather than the raw string.

For example, at Facebook if we return raw HTML from the server to render in React, we actually wrap it in the `__html` object on the server side, and return it as part of the response. This means the client-side won't get confused about which HTML blobs are save vs which HTML blobs are unsafe - If it's a `__html` object, it's safe.